### ISSUE TITLE
refactor(mlx-cache): cache 戦略統合と test isolation の改善

### ISSUE THREAD

**Issue by thkt** [thkt/rurico#101]
## 背景

rurico には複数のプロセスグローバル singleton と cache 戦略が存在し、それぞれ独立に進化している。bucket batching の前提と cache 設計の不整合、failure 永続化、test 環境での reset 不能、kernel recompile を誘発する cache-clear 戦略など、合計 6 つの finding が同領域に集まっている。

それぞれの優先度は異なり、一部は明確な bug、一部は intentional declared design、一部は test debt。本 issue は scope を「cache 戦略の整理」に絞り、明確な bug は本 PR で修正、intentional な design 変更は要件次第で別 PR / 別 issue に分ける。

### 1. local_mask_cache が single-entry（明確な mismatch）

```rust
// src/modernbert/model.rs:205-206
/// Single-entry cache keyed on `seq_len`. Rebuilds when `seq_len` changes.
local_mask_cache: Option<(i32, Array)>,
```

Phase 2 (bucket batching) で `seq_len` は bucket 数 (4) のいずれかに固定される設計だが、cache は 1 entry のみ。bucket 切替で必ず eviction、`get_local_attention_mask` が呼ばれるたび `seq_len × seq_len × 4 byte` の Array アロケーション。最大 8192² × 4 = 256MB / miss。

→ **明確な bug、本 PR で修正**

### 2. clear_inference_cache が batch ごとに呼ばれる（intentional declared）

```rust
// src/mlx_cache.rs:23-24
/// Clears both the buffer pool (mlx_clear_cache) and the compiled Metal
/// kernel cache (mlx_detail_compile_clear_cache). The compile cache grows
/// with each unique (batch_size, seq_len) pair; clearing it after every
/// batch prevents Metal OOM across long embedding runs.
```

bucket batching で unique shape は 4 種類に bounded されているため、毎バッチクリアは過剰可能。ただし comment が intentional design と明示している。kernel recompile の latency は production observe なし。

→ **本 PR では変更しない**。Metal OOM の上限調査と recompile の観測が前提

### 3. ensure_sqlite_vec の failure が永続化（intentional declared）

```rust
// src/storage.rs:36-39
return Err(format!(
    "sqlite-vec extension failed to register (sqlite3 rc={rc}). \
     This is a process-level initialization error that cannot be retried."
));
```

OnceLock により Err が永続化、process restart まで recover 不能。コメントは intentional declared。

→ **本 PR では変更しない**。retry path 必要性の検討は別 issue で

### 4. ensure_sqlite_vec failure に tracing event なし

`Err` は呼び出し元に返るが、`tracing::error!` 等の構造化イベントは emit されない。observability の観点では追加余地あり。

→ **本 PR で 1 行追加**

### 5. MLX_CACHE_LOCK / OnceLock の reset 不能

```rust
// src/mlx_cache.rs:14
pub(crate) static MLX_CACHE_LOCK: Mutex<()> = Mutex::new(());

// src/storage.rs:30
static INIT: OnceLock<Result<(), i32>> = OnceLock::new();
```

test で reset したい場面は現状観測されていない（既存 test は素直に動作している）。

→ **本 PR では変更しない**。test debt は個別 test 追加時に再評価

## アウトカム影響

| 項目                                | 観測可能 | 本 PR 対応                                                |
| ----------------------------------- | -------- | --------------------------------------------------------- |
| local_mask_cache single-entry       | yes      | yes — 4-entry HashMap に変更                              |
| clear_inference_cache 戦略          | -        | no — intentional declared、observe してから別 issue       |
| sqlite-vec failure 永続化           | -        | no — intentional declared、retry 必要性は別 issue         |
| sqlite-vec failure tracing 不在     | minor    | yes — `tracing::error!` 1 行追加                          |
| MLX_CACHE_LOCK / OnceLock reset 不能 | -       | no — test debt、現状 reset 必要なケースなし               |

### 1. local_mask_cache を multi-entry に

```rust
// src/modernbert/model.rs:195-207
pub struct ModernBert {
    // ...
    /// Per-seq-len cache. Bucket batching uses up to 4 unique seq_len values,
    /// so this map is bounded by the bucket count.
    local_mask_cache: HashMap<i32, Array>,
}
```

`get_local_attention_mask` の caller 側で：

```rust
let mask = self.local_mask_cache.entry(seq_len).or_insert_with(|| {
    get_local_attention_mask(seq_len, half_window).expect("validate-bound")
});
```

`expect` は別 RC (RC-007) で扱われる。本 PR は cache 構造のみ変更。

### 2. ensure_sqlite_vec failure に tracing イベント追加

```rust
// src/storage.rs:31-34
let init_result = INIT.get_or_init(|| {
    let rc = rurico_ffi::sqlite_vec_register();
    if rc == 0 {
        Ok(())
    } else {
        tracing::error!(rc, \"sqlite-vec auto-extension registration failed\");
        Err(rc)
    }
});
```

## 受け入れ条件

- `src/modernbert/model.rs` の `local_mask_cache` が `HashMap<i32, Array>` に変更
- bucket batching での 4 unique seq_len が同一 batch 内で全て cached、eviction なし
- ベンチマークまたは test で 4 種類 seq_len 入力時に各 mask が一度だけ allocate されることを確認
- `ensure_sqlite_vec` failure 時に `tracing::error!(rc, ...)` イベント emit
- `cargo test --workspace` 全 pass、clippy/fmt clean
- (任意) `docs/benchmarks/` に local_mask_cache 改善の前後比較を追記

## 解決対象（Wave 1 findings, /audit 2026-05-01-110240）

- RES-007 (`src/modernbert/model.rs:205` local_mask_cache single-entry eviction)
- PRF-006 (同箇所、performance 軸)
- OPS-013 (`src/storage.rs:30` sqlite-vec failure tracing 不在)

## 切り離し対象（別 Issue / defer）

- PRF-003 (`src/mlx_cache.rs:49-62` clear_inference_cache 戦略) — intentional declared、Metal OOM 観測前提なら別 issue
- RES-006 (`src/storage.rs:30` OnceLock failure 永続化) — intentional declared、retry 必要性検討は別 issue
- TST-006 (`src/mlx_cache.rs:14` MLX_CACHE_LOCK reset 不能) — test debt、reset 必要 test 出現時に再評価
- TST-007 (`src/storage.rs:30` OnceLock reset 不能) — 同上

## downstream 追従 (amici / yomu / recall / sae)

API 影響なし（内部実装 / docs / CI のみ）。merge 後、4 リポで `Cargo.toml` の `rurico` rev bump のみで取り込み完了。

- [ ] merge 後、amici / yomu / recall / sae 各リポに rev bump 用追従 issue を作成（既存 rurico 追従 issue があれば統合）

## 工数

実装 30min（local_mask_cache HashMap 化）+ tracing 5min + テスト 15min。レビュー別。
